u_fingerprinter.go

```diff
@@ -0,0 +1,330 @@
+				}
+				supportedSignatureAlgorithms = append(
+					supportedSignatureAlgorithms, SignatureScheme(sigAndAlg))
+			}
+			clientHelloSpec.Extensions = append(clientHelloSpec.Extensions, &SignatureAlgorithmsExtension{supportedSignatureAlgorithms})
+
+		case extensionSignatureAlgorithmsCert:
+			// RFC 8446, Section 4.2.3
+			if f.AllowBluntMimicry {
+				clientHelloSpec.Extensions = append(clientHelloSpec.Extensions, &GenericExtension{extension, extData})
+			} else {
+				return nil, errors.New("unsupported extension SignatureAlgorithmsCert")
+			}
+
+		case extensionRenegotiationInfo:
+			// RFC 5746, Section 3.2
+			clientHelloSpec.Extensions = append(clientHelloSpec.Extensions, &RenegotiationInfoExtension{RenegotiateOnceAsClient})
+
+		case extensionALPN:
+			// RFC 7301, Section 3.1
+			var protoList cryptobyte.String
+			if !extData.ReadUint16LengthPrefixed(&protoList) || protoList.Empty() {
+				return nil, errors.New("unable to read ALPN extension data")
+			}
+			alpnProtocols := []string{}
+			for !protoList.Empty() {
+				var proto cryptobyte.String
+				if !protoList.ReadUint8LengthPrefixed(&proto) || proto.Empty() {
+					return nil, errors.New("unable to read ALPN extension data")
+				}
+				alpnProtocols = append(alpnProtocols, string(proto))
+
+			}
+			clientHelloSpec.Extensions = append(clientHelloSpec.Extensions, &ALPNExtension{alpnProtocols})
+
+		case extensionSCT:
+			// RFC 6962, Section 3.3.1
+			clientHelloSpec.Extensions = append(clientHelloSpec.Extensions, &SCTExtension{})
+
+		case extensionSupportedVersions:
+			// RFC 8446, Section 4.2.1
+			var versList cryptobyte.String
+			if !extData.ReadUint8LengthPrefixed(&versList) || versList.Empty() {
+				return nil, errors.New("unable to read supported versions extension data")
+			}
+			supportedVersions := []uint16{}
+			for !versList.Empty() {
+				var vers uint16
+				if !versList.ReadUint16(&vers) {
+					return nil, errors.New("unable to read supported versions extension data")
+				}
+				supportedVersions = append(supportedVersions, unGREASEUint16(vers))
+			}
+			clientHelloSpec.Extensions = append(clientHelloSpec.Extensions, &SupportedVersionsExtension{supportedVersions})
+
+		case extensionKeyShare:
+			// RFC 8446, Section 4.2.8
+			var clientShares cryptobyte.String
+			if !extData.ReadUint16LengthPrefixed(&clientShares) {
+				return nil, errors.New("unable to read key share extension data")
+			}
+			keyShares := []KeyShare{}
+			for !clientShares.Empty() {
+				var ks KeyShare
+				var group uint16
+				if !clientShares.ReadUint16(&group) ||
+					!readUint16LengthPrefixed(&clientShares, &ks.Data) ||
+					len(ks.Data) == 0 {
+					return nil, errors.New("unable to read key share extension data")
+				}
+				ks.Group = CurveID(unGREASEUint16(group))
+				// if not GREASE, key share data will be discarded
+				// as it should be generated per connection
+				if ks.Group != GREASE_PLACEHOLDER {
+					ks.Data = nil
+				}
```

Comment:
Just a heads up that a teammate pointed out that the actual key_share "data" should be generated per connection. I could make another option for this, but it seems unlikely to actually be what the end user wants?

Reply:
I think users would appreciate this option, and don't you want this in your software? Also, have you had a chance to try integrating this feature? That may reveal existing issues, and allow to make changes to the API before it's committed.

I suspect that the option we'd want would be along the lines of DisableSecretsRegeneration, such that by default it would re-randomize ClientRandom and regenerate the keys in key_share, but one can set it to true to disable that.

Reply:
:+1: thanks for the point here. I don't think we're currently planning on needing it, but I did add a `DisableSecretsRegeneration` flag for future flexibility/options.

I've also added a `AlwaysAddPadding` option at the request of one of my colleagues.

Reply:
>  Also, have you had a chance to try integrating this feature? That may reveal existing issues, and allow to make changes to the API before it's committed.

We're very much in the process of this :grin: I _thiiink_ we may have it in production by next week, at least for some percentage of users.

So far the API suggestions you've made have been very helpful!

Reply:
Oh, my bad. When reading your first comment, I misunderstood you and thought that you weren't actually planning on regenerating the key_share data. Of course, regenerating it is the right behavior, I can't imagine why anyone wouldn't want to do it.
I'd even argue we should remove the DisableSecretsRegeneration. If someone really wants to do it, they actually can set whatever keys_share they want, it just would take some extra steps. I was only suggesting we add that option because I thought you wanted to not regenerate the secret for some reason, and wanted to give you the privilege of convenient API for contributing the feature.

Reply:
Gotcha - I've removed it now.